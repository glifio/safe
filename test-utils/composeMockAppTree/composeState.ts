import { FilecoinNumber } from '@glif/filecoin-number'
import { createPath, WalletProviderState } from '@glif/react-components'

import { IMPORT_MNEMONIC, IMPORT_SINGLE_KEY } from '../../constants'
import { mockWalletProviderInstance } from '../../__mocks__/@glif/filecoin-wallet-provider'
import { emptyMsigState } from '../../MsigProvider/types'
import {
  WALLET_ADDRESS,
  MULTISIG_ACTOR_ADDRESS,
  signers,
  WALLET_ID,
  WALLET_ADDRESS_2,
  WALLET_ID_2
} from '../constants'

export type Preset =
  | 'preOnboard'
  | 'postOnboard'
  | 'postOnboardLowBal'
  | 'postOnboardWithError'
  | 'selectedOtherWallet'
  | 'pendingMsigCreate'

export const composeWalletProviderState = (
  initialWalletProviderState: WalletProviderState,
  preset: Preset
) => {
  switch (preset) {
    case 'pendingMsigCreate':
    case 'postOnboard': {
      return Object.freeze({
        ...initialWalletProviderState,
        walletType: IMPORT_MNEMONIC,
        walletProvider: mockWalletProviderInstance,
        wallets: [
          {
            robust: WALLET_ADDRESS,
            id: WALLET_ID,
            balance: new FilecoinNumber('1', 'fil'),
            path: createPath(1, 0)
          }
        ],
        selectedWalletIdx: 0,
        loginOption: IMPORT_SINGLE_KEY
      })
    }
    case 'postOnboardLowBal': {
      return Object.freeze({
        ...initialWalletProviderState,
        walletType: IMPORT_MNEMONIC,
        walletProvider: mockWalletProviderInstance,
        wallets: [
          {
            robust: WALLET_ADDRESS,
            id: WALLET_ID,
            balance: new FilecoinNumber('.000001', 'fil'),
            path: createPath(1, 0)
          }
        ],
        selectedWalletIdx: 0,
        loginOption: IMPORT_SINGLE_KEY
      })
    }
    case 'postOnboardWithError': {
      return Object.freeze({
        ...initialWalletProviderState,
        walletProvider: mockWalletProviderInstance,
        wallets: [
          {
            robust: WALLET_ADDRESS,
            id: WALLET_ID,
            balance: new FilecoinNumber('1', 'fil'),
            path: createPath(1, 0)
          }
        ],
        selectedWalletIdx: 0,
        loginOption: IMPORT_SINGLE_KEY
      })
    }
    case 'selectedOtherWallet': {
      return Object.freeze({
        ...initialWalletProviderState,
        walletProvider: mockWalletProviderInstance,
        wallets: [
          {
            robust: WALLET_ADDRESS,
            id: WALLET_ID,
            balance: new FilecoinNumber('1', 'fil'),
            path: createPath(1, 0)
          },
          {
            robust: WALLET_ADDRESS_2,
            id: WALLET_ID_2,
            balance: new FilecoinNumber('5', 'fil'),
            path: createPath(1, 1)
          }
        ],
        selectedWalletIdx: 1,
        loginOption: IMPORT_MNEMONIC
      })
    }
    default:
      return initialWalletProviderState
  }
}

export const mockMsigProviderContext = {
  Address: MULTISIG_ACTOR_ADDRESS,
  ActorCode: 'fil/5/multisig',
  Balance: new FilecoinNumber('1', 'fil'),
  AvailableBalance: new FilecoinNumber('1', 'fil'),
  Signers: signers,
  InitialBalance: new FilecoinNumber('1', 'fil'),
  NextTxnID: 0,
  NumApprovalsThreshold: 2,
  StartEpoch: 0,
  UnlockDuration: 0,
  errors: {
    notMsigActor: false,
    connectedWalletNotMsigSigner: false,
    actorNotFound: false,
    unhandledError: ''
  },
  loading: false,
  setMsigActor: null
}

export const composeMsigProviderState = (preset: Preset) => {
  switch (preset) {
    case 'preOnboard': {
      return Object.freeze(emptyMsigState)
    }
    case 'pendingMsigCreate': {
      return Object.freeze(emptyMsigState)
    }
    default: {
      return Object.freeze(mockMsigProviderContext)
    }
  }
}
