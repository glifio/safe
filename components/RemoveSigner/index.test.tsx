import {
  render,
  act,
  waitFor,
  fireEvent,
  getByText,
  getByRole,
  RenderResult
} from '@testing-library/react'
import { Context } from 'react'
import { BigNumber } from '@glif/filecoin-number'
import { Message } from '@glif/filecoin-message'
import {
  MsigMethod,
  truncateAddress,
  WalletProviderContextType
} from '@glif/react-components'

import {
  pushPendingMessageSpy,
  WalletProviderContext,
  PendingMsgContext
} from '../../__mocks__/@glif/react-components'
import composeMockAppTree from '../../test-utils/composeMockAppTree'
import {
  flushPromises,
  WALLET_ADDRESS,
  MULTISIG_ACTOR_ADDRESS,
  MULTISIG_SIGNER_ADDRESS_2
} from '../../test-utils'
import { RemoveSigner } from '.'

jest.mock('@glif/filecoin-wallet-provider')

describe('RemoveSigner', () => {
  test('it allows a user to remove a signer address', async () => {
    const { Tree, walletProvider } = composeMockAppTree('postOnboard')
    let result: RenderResult | null = null

    await act(async () => {
      result = render(
        <Tree>
          <RemoveSigner
            signerAddress={MULTISIG_SIGNER_ADDRESS_2}
            walletProviderOpts={{
              context:
                WalletProviderContext as unknown as Context<WalletProviderContextType>
            }}
            pendingMsgContext={PendingMsgContext}
          />
        </Tree>
      )

      await flushPromises()

      // Get HTML elements
      const header = getByRole(result.container, 'heading')
      const oldSigner = getByRole(result.container, 'combobox')
      const cancel = getByText(result.container, 'Cancel')
      const review = getByText(result.container, 'Review')

      // Check initial state
      expect(header).toHaveTextContent('Remove a signer')
      expect(oldSigner).toHaveDisplayValue(
        truncateAddress(MULTISIG_SIGNER_ADDRESS_2)
      )
      expect(cancel).toBeEnabled()
      expect(review).toBeEnabled()

      // Click review
      fireEvent.click(review)
      await flushPromises()

      // The total amount should show after getting the tx fee
      await waitFor(
        () => expect(getByText(result.container, 'Total')).toBeInTheDocument(),
        { timeout: 10000 }
      )

      // The tx fee info should now be shown
      const maxFeeRegex =
        /You will not pay more than [0-9.]+ FIL for this transaction/i
      expect(getByText(result.container, maxFeeRegex)).toBeInTheDocument()

      // The expert mode toggle should shown and be off
      const expertMode = getByRole(result.container, 'checkbox')
      expect(expertMode).toBeInTheDocument()
      expect(expertMode).not.toBeChecked()

      // The send button should now be available
      const send = getByText(result.container, 'Send')
      expect(send).toBeInTheDocument()
      expect(send).toBeEnabled()

      // Click send
      fireEvent.click(send)
      await flushPromises()
    })

    // Check wallet provider calls
    expect(walletProvider.getNonce).toHaveBeenCalled()
    expect(walletProvider.wallet.sign).toHaveBeenCalled()
    expect(walletProvider.simulateMessage).toHaveBeenCalled()
    expect(walletProvider.sendMessage).toHaveBeenCalled()

    // Check if sent message was properly formatted
    const lotusMessage = walletProvider.wallet.sign.mock.calls[0][1]
    const message = Message.fromLotusType(lotusMessage)
    expect(message.from).toBe(WALLET_ADDRESS)
    expect(message.to).toBe(MULTISIG_ACTOR_ADDRESS)
    expect(message.nonce).toBeGreaterThanOrEqual(0)
    expect(message.value).toBeInstanceOf(BigNumber)
    expect(message.value.isEqualTo(0)).toBe(true)
    expect(message.method).toBe(MsigMethod.PROPOSE)
    expect(typeof message.params).toBe('string')
    expect(message.params).toBeTruthy()
    expect(message.gasPremium).toBeInstanceOf(BigNumber)
    expect(message.gasPremium.isGreaterThan(0)).toBe(true)
    expect(message.gasFeeCap).toBeInstanceOf(BigNumber)
    expect(message.gasFeeCap.isGreaterThan(0)).toBe(true)
    expect(message.gasLimit).toBeGreaterThan(0)

    // Check if pending message was properly pushed
    const pendingMsg = pushPendingMessageSpy.mock.calls[0][0]
    expect(typeof pendingMsg.cid).toBe('string')
    expect(pendingMsg.cid).toBeTruthy()
    expect(pendingMsg.from.robust).toBe(WALLET_ADDRESS)
    expect(pendingMsg.to.robust).toBe(MULTISIG_ACTOR_ADDRESS)
    expect(pendingMsg.height).toBe('')
    expect(typeof pendingMsg.params).toBe('string')
    expect(pendingMsg.params).toBeTruthy()
    expect(Number(pendingMsg.nonce)).toBeGreaterThanOrEqual(0)
    expect(Number(pendingMsg.method)).toBe(MsigMethod.PROPOSE)
    expect(Number(pendingMsg.value)).toBe(0)
    expect(Number(pendingMsg.gasFeeCap)).toBeGreaterThan(0)
    expect(Number(pendingMsg.gasLimit)).toBeGreaterThan(0)
    expect(Number(pendingMsg.gasPremium)).toBeGreaterThan(0)

    // Check snapshot
    expect(result.container.firstChild).toMatchSnapshot()
  })

  test('it renders the initial state correctly', async () => {
    const { Tree } = composeMockAppTree('postOnboard')
    let result: RenderResult | null = null

    await act(async () => {
      result = render(
        <Tree>
          <RemoveSigner signerAddress={MULTISIG_SIGNER_ADDRESS_2} />
        </Tree>
      )
    })

    // Check snapshot
    expect(result.container.firstChild).toMatchSnapshot()
  })
})
