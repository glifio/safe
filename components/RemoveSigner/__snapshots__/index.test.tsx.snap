// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemoveSigner it allows a user to remove a signer address 1`] = `
.c16 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  white-space: nowrap;
  cursor: pointer;
  gap: 0.75rem;
  padding: 0.5em 0.75em;
  color: var(--black);
  border-width: 1px;
  border-style: solid;
  border-radius: 8px;
  border-color: var(--black);
  background: transparent;
  font-size: 1.25em;
  pointer-events: none;
  color: var(--gray-medium);
  border-color: var(--gray-medium);
  background: transparent;
}

.c16 svg path {
  stroke: var(--black);
}

.c16:hover {
  color: var(--purple-medium);
  border-color: var(--purple-medium);
}

.c16:hover svg path {
  stroke: var(--purple-medium);
}

.c16:active {
  color: var(--white) !important;
  border-color: var(--purple-medium) !important;
  background: var(--purple-medium) !important;
}

.c16:active svg path {
  stroke: var(--white) !important;
}

.c16 svg path {
  stroke: var(--gray-medium);
}

.c17 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  white-space: nowrap;
  cursor: pointer;
  gap: 0.75rem;
  padding: 0.5em 0.75em;
  color: var(--black);
  border-width: 1px;
  border-style: solid;
  border-radius: 8px;
  border-color: var(--black);
  background: transparent;
  font-size: 1.25em;
  color: var(--white);
  border-color: var(--green-medium);
  background: var(--green-medium);
  pointer-events: none;
  color: var(--gray-medium);
  border-color: var(--gray-medium);
  background: transparent;
}

.c17 svg path {
  stroke: var(--black);
}

.c17:hover {
  color: var(--purple-medium);
  border-color: var(--purple-medium);
}

.c17:hover svg path {
  stroke: var(--purple-medium);
}

.c17:active {
  color: var(--white) !important;
  border-color: var(--purple-medium) !important;
  background: var(--purple-medium) !important;
}

.c17:active svg path {
  stroke: var(--white) !important;
}

.c17 svg path {
  stroke: var(--white);
}

.c17:hover {
  color: var(--green-medium);
  border-color: var(--green-medium);
  background: var(--white);
}

.c17:hover svg path {
  stroke: var(--green-medium);
}

.c17 svg path {
  stroke: var(--gray-medium);
}

.c3 {
  display: inline-block;
  height: 48px;
  width: 48px;
  text-align: left;
  background-color: var(--purple-medium);
}

.c4 {
  display: inline-block;
  height: 24px;
  width: 24px;
  background-color: var(--purple-dark);
  -webkit-animation-name: bvGDKI;
  animation-name: bvGDKI;
  -webkit-animation-duration: 2s;
  animation-duration: 2s;
  -webkit-animation-iteration-count: infinite;
  animation-iteration-count: infinite;
}

.c1 {
  padding: 1.5em;
  border-radius: 8px;
  overflow: hidden;
  text-align: center;
  word-break: break-word;
}

.c1 > *:first-child {
  margin-top: 0;
}

.c1 > *:last-child {
  margin-bottom: 0;
}

.c1 > svg:first-child {
  display: block;
  margin: 0 auto 1em;
}

.c1 > h2:first-child {
  margin: 0;
}

.c1 header:first-child {
  margin: -1.5em -1.5em 1.5em -1.5em;
  padding: 1.5em;
  background-color: var(--purple-medium);
  color: var(--white);
}

.c2 {
  background-color: var(--blue-gray);
}

.c2 hr {
  border-color: var(--gray-medium);
}

.c5 {
  background-color: var(--white);
  box-shadow: 0 0 0.5em var(--gray-light);
}

.c14 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 10px;
  margin: 1.5em 0;
}

.c15 {
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 10px;
  width: 100%;
  max-width: 35em;
}

.c0 form {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 10px;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 10px;
}

.c6 > p,
.c6 > hr {
  margin: 0.5em 0;
}

.c6 > p[role='button'] {
  margin: 0;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 10px;
  text-align: left;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c10 > *:first-child {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c10 > *:last-child:not(.toggle-wrapper) {
  -webkit-flex: 0 0 50%;
  -ms-flex: 0 0 50%;
  flex: 0 0 50%;
}

.c10 span.info,
.c10 span.error {
  font-size: 0.875rem;
}

.c10 span.error {
  color: var(--red-medium);
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 10px;
  text-align: left;
  pointer-events: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c8 > *:first-child {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c8 > *:last-child:not(.toggle-wrapper) {
  -webkit-flex: 0 0 50%;
  -ms-flex: 0 0 50%;
  flex: 0 0 50%;
}

.c8 span.info,
.c8 span.error {
  font-size: 0.875rem;
}

.c8 span.error {
  color: var(--red-medium);
}

.c11 .info-wrapper {
  padding: 0.7em 1em;
  text-align: right;
}

.c9 .select-wrapper {
  position: relative;
}

.c9 .select-wrapper select {
  width: 100%;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c9 .select-wrapper .select-arrow {
  position: absolute;
  pointer-events: none;
  top: 50%;
  right: 1.2em;
  width: 0.6em;
  height: 0.6em;
  border-right: 1.5px solid var(--black);
  border-bottom: 1.5px solid var(--black);
  -webkit-transform: translateY(-0.35em) rotate(45deg);
  -ms-transform: translateY(-0.35em) rotate(45deg);
  transform: translateY(-0.35em) rotate(45deg);
  border-color: var(--gray-dark);
}

.c12 .toggle-wrapper {
  position: relative;
}

.c12 input {
  position: absolute;
  top: 50%;
  left: 50%;
  width: 0;
  height: 0;
  opacity: 0;
}

.c12 .toggle {
  display: inline-block;
  position: relative;
  width: 2.25em;
  height: 1.5em;
  border-radius: 0.75em;
  border: 2px solid var(--blue-medium);
  background-color: var(--blue-medium);
  border-color: var(--gray-light) !important;
  background-color: var(--gray-light) !important;
}

.c12 .toggle::after {
  content: '';
  position: absolute;
  top: 0;
  left: 0;
  width: calc(1.5em - 4px);
  height: calc(1.5em - 4px);
  border-radius: 50%;
  background-color: var(--white);
}

.c12 input:checked ~ .toggle {
  border-color: var(--blue-dark);
  background-color: var(--blue-dark);
}

.c12 input:checked ~ .toggle::after {
  left: auto;
  right: 0;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  gap: 20px;
}

.c7 > * {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
}

.c7 > *:first-child {
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
}

.c13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c13 span {
  color: var(--purple-medium);
  font-size: 1.75rem;
}

<div
  class="c0"
>
  <form>
    <div
      class="c1 c2"
    >
      <h2>
        Remove a signer
      </h2>
      <hr />
      <p>
        Sending your transaction...
      </p>
      <span
        class="c3"
      >
        <span
          class="c4"
        />
      </span>
    </div>
    
    <div
      class="c1 c5"
    >
      <div
        class="c6"
      >
        <header
          class="c7"
        >
          <div>
            <span>
              From
            </span>
            <span>
              t1z225 ... 6z6wgi
            </span>
          </div>
          <div>
            <span>
              Signer 
              Balance
            </span>
            <span>
              1
               FIL
            </span>
          </div>
          <div>
            <span>
              Safe Balance
            </span>
            <span>
              1
               FIL
            </span>
          </div>
        </header>
        <label
          class="c8 c9"
          disabled=""
        >
          <div>
            <span>
              Signer
            </span>
            
          </div>
          <div
            class="select-wrapper"
          >
            <select
              disabled=""
            >
              
              <option
                value="t0102"
              >
                t0102
              </option>
            </select>
            <div
              class="select-arrow"
            />
          </div>
        </label>
        <label
          class="c10 c11"
        >
          <div>
            <span>
              Required approvals
            </span>
            <span
              class="info"
            >
              The Safe will have 1 owners
            </span>
          </div>
          <div
            class="info-wrapper"
          >
            1
          </div>
        </label>
        <hr />
        <label
          class="c8 c12"
          disabled=""
        >
          <div>
            <span>
              Expert Mode
            </span>
            
          </div>
          <div
            class="toggle-wrapper"
          >
            <input
              disabled=""
              type="checkbox"
            />
            <span
              class="toggle"
            />
          </div>
        </label>
        <p>
          You will not pay more than 
          0.00000000001
           FIL for this transaction.
           
          <a
            href="https://filfox.info/en/stats/gas"
            rel="noreferrer noopener"
            target="_blank"
          >
            More information on average gas fee statistics.
          </a>
        </p>
        <hr />
        <div
          class="c13"
        >
          <span>
            Total
          </span>
          <span>
            0.00000000001
             FIL
          </span>
        </div>
      </div>
    </div>
    <div
      class="c14 c15"
    >
      <button
        class="c16"
        disabled=""
        type="button"
      >
        Back
      </button>
      <button
        class="c17"
        disabled=""
        type="submit"
      >
        Send
      </button>
    </div>
  </form>
</div>
`;

exports[`RemoveSigner it renders the initial state correctly 1`] = `
.c12 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  white-space: nowrap;
  cursor: pointer;
  gap: 0.75rem;
  padding: 0.5em 0.75em;
  color: var(--black);
  border-width: 1px;
  border-style: solid;
  border-radius: 8px;
  border-color: var(--black);
  background: transparent;
  font-size: 1.25em;
}

.c12 svg path {
  stroke: var(--black);
}

.c12:hover {
  color: var(--purple-medium);
  border-color: var(--purple-medium);
}

.c12:hover svg path {
  stroke: var(--purple-medium);
}

.c12:active {
  color: var(--white) !important;
  border-color: var(--purple-medium) !important;
  background: var(--purple-medium) !important;
}

.c12:active svg path {
  stroke: var(--white) !important;
}

.c13 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  white-space: nowrap;
  cursor: pointer;
  gap: 0.75rem;
  padding: 0.5em 0.75em;
  color: var(--black);
  border-width: 1px;
  border-style: solid;
  border-radius: 8px;
  border-color: var(--black);
  background: transparent;
  font-size: 1.25em;
  color: var(--white);
  border-color: var(--green-medium);
  background: var(--green-medium);
}

.c13 svg path {
  stroke: var(--black);
}

.c13:hover {
  color: var(--purple-medium);
  border-color: var(--purple-medium);
}

.c13:hover svg path {
  stroke: var(--purple-medium);
}

.c13:active {
  color: var(--white) !important;
  border-color: var(--purple-medium) !important;
  background: var(--purple-medium) !important;
}

.c13:active svg path {
  stroke: var(--white) !important;
}

.c13 svg path {
  stroke: var(--white);
}

.c13:hover {
  color: var(--green-medium);
  border-color: var(--green-medium);
  background: var(--white);
}

.c13:hover svg path {
  stroke: var(--green-medium);
}

.c1 {
  padding: 1.5em;
  border-radius: 8px;
  overflow: hidden;
  text-align: center;
  word-break: break-word;
}

.c1 > *:first-child {
  margin-top: 0;
}

.c1 > *:last-child {
  margin-bottom: 0;
}

.c1 > svg:first-child {
  display: block;
  margin: 0 auto 1em;
}

.c1 > h2:first-child {
  margin: 0;
}

.c1 header:first-child {
  margin: -1.5em -1.5em 1.5em -1.5em;
  padding: 1.5em;
  background-color: var(--purple-medium);
  color: var(--white);
}

.c2 {
  background-color: var(--blue-gray);
}

.c2 hr {
  border-color: var(--gray-medium);
}

.c3 {
  background-color: var(--white);
  box-shadow: 0 0 0.5em var(--gray-light);
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 10px;
  margin: 1.5em 0;
}

.c11 {
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 10px;
  width: 100%;
  max-width: 35em;
}

.c0 form {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 10px;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 10px;
}

.c4 > p,
.c4 > hr {
  margin: 0.5em 0;
}

.c4 > p[role='button'] {
  margin: 0;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 10px;
  text-align: left;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c6 > *:first-child {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c6 > *:last-child:not(.toggle-wrapper) {
  -webkit-flex: 0 0 50%;
  -ms-flex: 0 0 50%;
  flex: 0 0 50%;
}

.c6 span.info,
.c6 span.error {
  font-size: 0.875rem;
}

.c6 span.error {
  color: var(--red-medium);
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 10px;
  text-align: left;
  pointer-events: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c8 > *:first-child {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c8 > *:last-child:not(.toggle-wrapper) {
  -webkit-flex: 0 0 50%;
  -ms-flex: 0 0 50%;
  flex: 0 0 50%;
}

.c8 span.info,
.c8 span.error {
  font-size: 0.875rem;
}

.c8 span.error {
  color: var(--red-medium);
}

.c7 .select-wrapper {
  position: relative;
}

.c7 .select-wrapper select {
  width: 100%;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c7 .select-wrapper .select-arrow {
  position: absolute;
  pointer-events: none;
  top: 50%;
  right: 1.2em;
  width: 0.6em;
  height: 0.6em;
  border-right: 1.5px solid var(--black);
  border-bottom: 1.5px solid var(--black);
  -webkit-transform: translateY(-0.35em) rotate(45deg);
  -ms-transform: translateY(-0.35em) rotate(45deg);
  transform: translateY(-0.35em) rotate(45deg);
}

.c9 .toggle-wrapper {
  position: relative;
}

.c9 input {
  position: absolute;
  top: 50%;
  left: 50%;
  width: 0;
  height: 0;
  opacity: 0;
}

.c9 .toggle {
  display: inline-block;
  position: relative;
  width: 2.25em;
  height: 1.5em;
  border-radius: 0.75em;
  border: 2px solid var(--blue-medium);
  background-color: var(--blue-medium);
  border-color: var(--gray-light) !important;
  background-color: var(--gray-light) !important;
}

.c9 .toggle::after {
  content: '';
  position: absolute;
  top: 0;
  left: 0;
  width: calc(1.5em - 4px);
  height: calc(1.5em - 4px);
  border-radius: 50%;
  background-color: var(--white);
}

.c9 input:checked ~ .toggle {
  border-color: var(--blue-dark);
  background-color: var(--blue-dark);
}

.c9 input:checked ~ .toggle::after {
  left: auto;
  right: 0;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  gap: 20px;
}

.c5 > * {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
}

.c5 > *:first-child {
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
}

<div
  class="c0"
>
  <form>
    <div
      class="c1 c2"
    >
      <h2>
        Remove a signer
      </h2>
      <hr />
      <p>
        Please select the signer address to remove below
      </p>
    </div>
    
    <div
      class="c1 c3"
    >
      <div
        class="c4"
      >
        <header
          class="c5"
        >
          <div>
            <span>
              From
            </span>
            <span>
              t1z225 ... 6z6wgi
            </span>
          </div>
          <div>
            <span>
              Signer 
              Balance
            </span>
            <span>
              1
               FIL
            </span>
          </div>
          <div>
            <span>
              Safe Balance
            </span>
            <span>
              1
               FIL
            </span>
          </div>
        </header>
        <label
          class="c6 c7"
        >
          <div>
            <span>
              Signer
            </span>
            
          </div>
          <div
            class="select-wrapper"
          >
            <select>
              
              <option
                value="t0102"
              >
                t0102
              </option>
            </select>
            <div
              class="select-arrow"
            />
          </div>
        </label>
        <label
          class="c8 c9"
          disabled=""
        >
          <div>
            <span>
              Decrease required approvals
            </span>
            <span
              class="info"
            >
              From 2 to 1
            </span>
          </div>
          <div
            class="toggle-wrapper"
          >
            <input
              checked=""
              disabled=""
              type="checkbox"
            />
            <span
              class="toggle"
            />
          </div>
        </label>
      </div>
    </div>
    <div
      class="c10 c11"
    >
      <button
        class="c12"
        type="button"
      >
        Cancel
      </button>
      <button
        class="c13"
        type="submit"
      >
        Review
      </button>
    </div>
  </form>
</div>
`;
